lib/download.go

```diff
@@ -1,47 +1,137 @@
+	pubKeyFilename := filepath.Join(installLocation, "/", PubKeyPrefix+PubKeyId+pubKeySuffix)
+	zipUrl := mirrorURL + tfversion + "/" + versionPrefix + tfversion + "_" + goos + "_" + goarch + ".zip"
+	hashUrl := mirrorURL + tfversion + "/" + versionPrefix + tfversion + "_SHA256SUMS"
+	hashSignatureUrl := mirrorURL + tfversion + "/" + versionPrefix + tfversion + "_SHA256SUMS." + PubKeyId + ".sig"
+
+	err := downloadPublicKey(installLocation, pubKeyFilename)
+	if err != nil {
+		log.Fatal("[Error]: Could not download public key file")
+	}
+
+	log.Println("Downloading ", zipUrl)
+	zipFilePath, err := downloadFromURL(installLocation, zipUrl)
+	if err != nil {
+		log.Fatal("[Error]: Could not download zip file")
+	}
+
+	log.Println("Downloading ", hashUrl)
+	hashFilePath, err := downloadFromURL(installLocation, hashUrl)
+	if err != nil {
+		log.Fatal("[Error]: Could not download hash file")
+	}
+
+	log.Println("Downloading ", hashSignatureUrl)
+	hashSigFilePath, err := downloadFromURL(installLocation, hashSignatureUrl)
+	if err != nil {
+		log.Fatal("[Error]: Could not download hash signature file")
+	}
+
+	publicKeyFile, err := os.Open(pubKeyFilename)
+	if err != nil {
+		log.Fatal("[Error]: Could not open the public key.", pubKeyFilename)
+	}
+
+	signatureFile, err := os.Open(hashSigFilePath)
+	if err != nil {
+		log.Fatal("[Error]: Could not open the signature file.", hashSigFilePath)
+	}
+
+	targetFile, err := os.Open(zipFilePath)
+	if err != nil {
+		log.Fatal("[Error]: Could not open the terraform binary for checksum verification.", zipFilePath)
+	}
+
+	hashFile, err := os.Open(hashFilePath)
+	if err != nil {
+		log.Fatal("[Error]: Could not open the terraform checksum file for signature verification.", hashFilePath)
+	}
+	verified := checkSignatureOfChecksums(publicKeyFile, hashFile, signatureFile)
+	if !verified {
+		return "", errors.New("signature of checksum files could not be verified")
+	}
+	match := checkChecksumMatches(hashFilePath, targetFile)
+	if !match {
+		return "", errors.New("checksums did not match")
+	}
+	return zipFilePath, err
+}
+
+func downloadFromURL(installLocation string, url string) (string, error) {
 	tokens := strings.Split(url, "/")
 	fileName := tokens[len(tokens)-1]
-	fmt.Printf("Downloading to: %s\n", installLocation)
+	log.Printf("Downloading to: %s\n", filepath.Join(installLocation, "/", fileName))
 
 	response, err := http.Get(url)
 	if err != nil {
-		fmt.Println("[Error] : Error while downloading", url, "-", err)
+		log.Fatal("[Error] : Error while downloading", url, "-", err)
 		return "", err
 	}
 	defer response.Body.Close()
 
 	if response.StatusCode != 200 {
 		//Sometimes hashicorp terraform file names are not consistent
 		//For example 0.12.0-alpha4 naming convention in the release repo is not consistent
-		return "", fmt.Errorf("[Error] : Unable to download from %s", url)
+		log.Fatalf("[Error] : Unable to download from %s", url)
 	}
 
 	zipFile := filepath.Join(installLocation, fileName)
```

Comment:
If this function is downloading arbitrary files now, zipFile probably isn't too accurate.

Reply:
Fixed